### PR TITLE
OTLP: ignore duplicate metric data points

### PR DESCRIPTION
**Pull request by felixbarny**
Part of https://github.com/elastic/elasticsearch/pull/133057

Handles the case where a metric data point with a duplicate metric name is added to a `DataPointGroup`.

### CODE REVIEW

File: x-pack/plugin/otel-data/src/main/java/org/elasticsearch/xpack/oteldata/otlp/datapoint/DataPointGroupingContext.java [hunk 1]
```diff
@@ -276,9 +277,17 @@ public String getMetricNamesHash() {
             return metricNamesHash;
         }
 
-        public void addDataPoint(DataPoint dataPoint) {
+        boolean addDataPoint(Set<String> ignoredDataPointMessages, DataPoint dataPoint) {
             metricNamesHash = null; // reset the hash when adding a new data point
-            dataPoints.add(dataPoint);
+            if (metricNames.add(dataPoint.getMetricName()) == false) {
+                ignoredDataPointMessages.add(
+                    "Duplicate metric name '" + dataPoint.getMetricName() + "' for timestamp " + getTimestampUnixNano()
```
Comment (kkrik-es) [inline x-pack/plugin/otel-data/src/main/java/org/elasticsearch/xpack/oteldata/otlp/datapoint/DataPointGroupingContext.java:284]: List dimensions too? Duplicate metrics should be allowed if the dimensions are different, iiuc?

Comment (felixbarny) [reply x-pack/plugin/otel-data/src/main/java/org/elasticsearch/xpack/oteldata/otlp/datapoint/DataPointGroupingContext.java:284]: > List dimensions too?

Hm, I'm a bit torn. On the one hand, it can make debugging this easier. On the other hand, this might make the error message quite verbose.

> Duplicate metrics should be allowed if the dimensions are different, iiuc?

Yes, that's right.

Comment (felixbarny) [reply x-pack/plugin/otel-data/src/main/java/org/elasticsearch/xpack/oteldata/otlp/datapoint/DataPointGroupingContext.java:284]: FWIW, in the equivalent error message that the `elasticsearchexporter` is producing, we're also not listing all dimensions: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b42536f0eaa6790e004be61767b1d2714e3f4ee3/exporter/elasticsearchexporter/internal/serializer/otelserializer/metrics.go#L60-L64

I suggest we consider adding an extended error message separate from this PR.

File: x-pack/plugin/otel-data/src/test/java/org/elasticsearch/xpack/oteldata/otlp/datapoint/DataPointGroupingContextTests.java [hunk 2]
```diff
@@ -74,6 +75,24 @@ public void testGroupingDifferentGroupUnit() throws Exception {
         assertEquals(2, groupCount.get());
     }
 
+    public void testGroupingDuplicateName() throws Exception {
```
Comment (kkrik-es) [inline x-pack/plugin/otel-data/src/test/java/org/elasticsearch/xpack/oteldata/otlp/datapoint/DataPointGroupingContextTests.java]: Let's add a test with duplicate name and different dimensions, to verify that it's accepted.

**Review by kkrik-es — APPROVED** [review]
